File: Kakao/2019_오픈채팅방.cpp
// O(record.length)

#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

struct Record {
    string command;
    string uid;
    Record(string c, string u):command(c),uid(u){}
};

vector<string> solution(vector<string> record) {
    vector<string> answer;
    vector<Record> record_parsed;
    unordered_map<string,string> uid_to_nickname;
    
    for(int i=0;i<record.size();i++) {
        int first = record[i].find_first_of(" ");
        int last = record[i].find_last_of(" ");
        string command = record[i].substr(0,first);
        string uid = record[i].substr(first+1,last-first-1);
        string nickname = record[i].substr(last+1);
        
        if(command=="Enter" || command=="Change")   uid_to_nickname[uid] = nickname;
        record_parsed.push_back(Record(command,uid));
    }
    
    for(int i=0;i<record_parsed.size();i++) {
        if(record_parsed[i].command=="Change") continue;
        
        string message;
        message += (uid_to_nickname[record_parsed[i].uid]+"님이 ");
        if(record_parsed[i].command=="Enter")  message += "들어왔습니다.";
        else    message += "나갔습니다.";
        answer.push_back(message);
    }
    return answer;
}

File: Kakao/2019_오픈채팅방.swift
// 출제의도: "Hash Table(Dictionary)", Time Complexity: O(L), Space Complexity: O(n). (L: record 모든 문자열 길이의 합)

import Foundation

func solution(_ record:[String]) -> [String] {
    var nickname = [String: String]()

    record.forEach {
        let seperated = $0.components(separatedBy: " ")
        if seperated.count==3 {
            nickname[seperated[1]] = seperated[2]
        }
    }

    return record.filter({!$0.contains("Change")})
        .map({ (log:String) -> String in
            let seperated = log.components(separatedBy: " ")
            var message = nickname[seperated[1]]!

            switch(seperated[0]) {
                case "Enter":
                    message += "님이 들어왔습니다."
                case "Leave":
                    message += "님이 나갔습니다."
                default:
                    break
            }
              
            return message
      })
}
